### PR TITLE
Automate and manual tests for label-driven recurring job

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/467

### CODE REVIEW

File: manager/integration/tests/common.py [hunk 1]
```diff
@@ -4039,6 +4080,27 @@ def backing_image_feature_supported(client):
...
 
+def cleanup_all_recurring_jobs(client):
+    recurring_jobs = client.list_recurring_job()
+    for recurring_job in recurring_jobs:
+        try:
+            client.delete(recurring_job)
+        except Exception as e:
+            print("\nException when cleanup recurring job ", recurring_job)
+            print(e)
+            pass
```
Comment (innobead) [inline manager/integration/tests/common.py:4091]: Pass should be needless.

File: manager/integration/tests/common.py [hunk 2]
```diff
@@ -4039,6 +4080,27 @@ def backing_image_feature_supported(client):
...
 
 
+def cleanup_all_recurring_jobs(client):
+    recurring_jobs = client.list_recurring_job()
+    for recurring_job in recurring_jobs:
+        try:
+            client.delete(recurring_job)
+        except Exception as e:
+            print("\nException when cleanup recurring job ", recurring_job)
+            print(e)
```
Comment (innobead) [inline manager/integration/tests/common.py:4090]: Minor, but u can merge two lines as one print statement.

File: manager/integration/tests/common.py [hunk 3]
```diff
@@ -4052,3 +4114,111 @@ def get_engine_image_status_value(client, ei_name):
...
+                                    Cron=spec["cron"],
+                                    Retain=spec["retain"],
+                                    Concurrency=spec["concurrency"])
+        policy = client.by_id_recurring_job(name)
+        assert policy.name == name
+        assert policy.task == spec["task"]
+        assert policy.groups == spec["groups"]
+        assert policy.cron == spec["cron"]
+        assert policy.retain == spec["retain"]
+        assert policy.concurrency == spec["concurrency"]
```
Comment (innobead) [inline manager/integration/tests/common.py:4133]: Assert here could be optional in common libs but usually used in test cases condition check instead.

File: manager/integration/tests/test_recurring_job.py [hunk 4]
```diff
@@ -759,3 +794,641 @@ def backup_complete_predicate(b):
...
+    # wait_until_begin_of_a_minute()
+    common.write_volume_random_data(volume1)
+    common.write_volume_random_data(volume2)
+    time.sleep(60)
+    wait_for_snapshot_count(volume1, 3)
+    wait_for_snapshot_count(volume2, 3)
+
+
+@pytest.mark.recurring_job  # NOQA
+def test_recurring_job_policy_backup(set_random_backupstore, client, batch_v1_beta_api):  # NOQA
```
Comment (innobead) [inline manager/integration/tests/test_recurring_job.py:1332]: Probably we can remove policy keyword

File: manager/integration/tests/test_recurring_job.py [hunk 5]
```diff
@@ -759,3 +794,641 @@ def backup_complete_predicate(b):
...
+            TASK: BACKUP,
+            GROUPS: [group1],
+            CRON: "* * * * *",
+            RETAIN: 1,
+            CONCURRENCY: 2
+        },
+    }
+    create_recurring_jobs(client, recurring_jobs)
+
+    volume1.recurringJobAdd(name=group1, isGroup=True)
```
Comment (innobead) [inline manager/integration/tests/test_recurring_job.py:843]: Python method name should be snake case

**Review by innobead — COMMENTED** [review]
In general, LGTM. Some existing recurring job test cases needed to modify as u mentioned by TODO.

File: manager/integration/tests/test_recurring_job.py (hunk 5 above)
Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py:843]: This POST queried for [actions](https://github.com/longhorn/longhorn-manager/blob/master/api/router.go#L68-L95).

File: manager/integration/tests/common.py (hunk 2 above)
Comment (c3y1huang) [reply manager/integration/tests/common.py:4090]: Without going over the line character's limit, I think the error would be easier to read on a separate line. And this is currently consistent with the others.

Comment (innobead) [reply manager/integration/tests/common.py:4090]: I mean `print("....", var, e)` or even use .format. But that's fine if most of the code has been done with this style already.

File: manager/integration/tests/test_recurring_job.py (hunk 5 above)
Comment (innobead) [reply manager/integration/tests/test_recurring_job.py:843]: Thanks :), although this does not follow up with the language convention, that's fine for now based on our own convention.

File: manager/integration/tests/common.py (hunk 1 above)
Comment (c3y1huang) [reply manager/integration/tests/common.py]: Removed.

File: manager/integration/tests/common.py (hunk 3 above)
Comment (c3y1huang) [reply manager/integration/tests/common.py:4134]: Make sense. Splited the checking to another function.

File: manager/integration/tests/test_recurring_job.py (hunk 4 above)
Comment (c3y1huang) [reply manager/integration/tests/test_recurring_job.py]: Oops. Thanks, I've missed a few... fixed.

**Comment by c3y1huang** [comment]
Regression Build - [398](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/398/)

**Comment by c3y1huang** [comment]
Regression - [407](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/407/testReport/)

File: manager/integration/tests/test_recurring_job.py [hunk 6]
```diff
@@ -4,60 +4,94 @@
...
+RETAIN = "retain"
+SNAPSHOT = "snapshot"
+BACKUP = "backup"
+CONCURRENCY = "concurrency"
+LABELS = "labels"
+DEFAULT = "default"
+SCHEDULE_1MIN = "* * * * *"
+
+
+def wait_until_begin_of_a_minute():
```
Comment (innobead) [inline manager/integration/tests/test_recurring_job.py:89]: Why do we need this?

File: manager/integration/tests/test_recurring_job.py [hunk 7]
```diff
@@ -658,103 +951,619 @@ def test_recurring_jobs_on_nodes_with_taints():  # NOQA
...
+
+    Given volume `test-job-1` created, attached and healthy.
+    And create `backup1`   recurring job with `default` in groups.
+        create `backup2`   recurring job with ``        in groups.
+    And `default` group exist in `test-job-1` volume recurring job label.
+    And `backup1` cron job exist.
+        `backup2` cron job exist.
+    And write data to  `test-job-1` volume.
+    And 2 snapshot exist in `test-job-1` volume.
+    And 1 backup   exist in `test-job-2` volume.
```
Comment (innobead) [inline manager/integration/tests/test_recurring_job.py:1354]: There is no `given` to describe `test-job-2`? why we have this `And`?

File: manager/integration/tests/test_recurring_job.py [hunk 8]
```diff
@@ -658,103 +951,619 @@ def test_recurring_jobs_on_nodes_with_taints():  # NOQA
...
+
+    When create and attach volume `test-job-2`.
+         wait for volume `test-job-2` to be healthy.
+    And `default` group exist in `test-job-2` volume recurring job label.
+    And write data to  `test-job-1` volume.
+    Then 2 snapshot exist in `test-job-1` volume.
+         1 backup   exist in `test-job-2` volume.
```
Comment (innobead) [inline manager/integration/tests/test_recurring_job.py:1361]: These should be for `test-job-2`

**Review by innobead — APPROVED** [review]
In general, LGTM. Merging to resolve for the change of code change. 

P.S. some test description maybe needs to adjust, but @c3y1huang can update it later.

cc @khushboo-rancher @meldafrawi @kaxing

**Pull request merged**